tests/integration/cli/tests/run.rs

```diff
@@ -508,6 +508,115 @@ fn run_no_imports_wasm_works() -> anyhow::Result<()> {
+
+    let stderr_lines = stderr.lines().map(|s| s.to_string()).collect::<Vec<_>>();
+
+    assert_eq!(
+        stderr_lines,
+        vec!["error: invalid package name, could not find file does/not/exist".to_string()]
+    );
+
+    Ok(())
+}
+
+#[test]
+fn run_test_caching_works_for_packages() -> anyhow::Result<()> {
+    let output = Command::new(get_wasmer_path())
+        .arg("python/python")
+        .arg(format!("--mapdir=.:{}", ASSET_PATH))
+        .arg("test.py")
+        .output()?;
+
+    if output.stdout != b"hello\n".to_vec() {
+        panic!("failed to run https://wapm.io/python/python for the first time");
+    }
+
+    let time = std::time::Instant::now();
+
+    let output = Command::new(get_wasmer_path())
+        .arg("python/python")
+        .arg(format!("--mapdir=.:{}", ASSET_PATH))
+        .arg("test.py")
+        .output()?;
+
+    if output.stdout != b"hello\n".to_vec() {
+        panic!("failed to run https://wapm.io/python/python for the second time");
+    }
+
+    // package should be cached
+    assert!(std::time::Instant::now() - time < std::time::Duration::from_secs(1));
+
+    Ok(())
+}
+
+#[test]
+fn run_test_caching_works_for_packages_with_versions() -> anyhow::Result<()> {
+    let output = Command::new(get_wasmer_path())
+        .arg("python/python@0.1.0")
+        .arg(format!("--mapdir=.:{}", ASSET_PATH))
+        .arg("test.py")
+        .output()?;
+
+    if output.stdout != b"hello\n".to_vec() {
+        panic!("failed to run https://wapm.io/python/python for the first time");
+    }
+
+    let time = std::time::Instant::now();
+
+    let output = Command::new(get_wasmer_path())
+        .arg("python/python@0.1.0")
+        .arg(format!("--mapdir=.:{}", ASSET_PATH))
+        .arg("test.py")
+        .output()?;
+
+    if output.stdout != b"hello\n".to_vec() {
+        panic!("failed to run https://wapm.io/python/python for the second time");
+    }
+
+    // package should be cached
+    assert!(std::time::Instant::now() - time < std::time::Duration::from_secs(1));
+
+    Ok(())
+}
+
+#[test]
+fn run_test_caching_works_for_urls() -> anyhow::Result<()> {
+    let output = Command::new(get_wasmer_path())
+        .arg("https://wapm.io/python/python")
+        .arg(format!("--mapdir=.:{}", ASSET_PATH))
+        .arg("test.py")
+        .output()?;
+
+    if output.stdout != b"hello\n".to_vec() {
+        panic!("failed to run https://wapm.io/python/python for the first time");
+    }
+
+    let time = std::time::Instant::now();
+
+    let output = Command::new(get_wasmer_path())
+        .arg("https://wapm.io/python/python")
+        .arg(format!("--mapdir=.:{}", ASSET_PATH))
+        .arg("test.py")
+        .output()?;
+
+    if output.stdout != b"hello\n".to_vec() {
+        panic!("failed to run https://wapm.io/python/python for the second time");
+    }
+
+    // package should be cached
+    assert!(std::time::Instant::now() - time < std::time::Duration::from_secs(1));
+
+    Ok(())
+}
+
```

Comment:
Needs tests for:

* Running local file
* Running with a url

Reply:
Running with a URL is already in the test?

Running with a local file is already tested in every other integration test (for example the `test_no_start_wat_path` runs the local `no_start.wat` file).

Reply:
Ah, right. I overlooked the `.arg("https://wapm.io/python/python")` test.